### PR TITLE
Add support for BindAsync without ParameterInfo

### LINKED ISSUE #36415: Support BindAsync() without ParameterInfo for Minimal APIs
### Is your feature request related to a problem? Please describe.

If a user wishes to use custom data binding for Minimal APIs using the `BindAsync()` pattern but does not require the `ParameterInfo` parameter when using Visual Studio, the IDE will warn about the unused `parameter` argument with `IDE0060`. I would _assume_ that other static analyzers would produce similar warnings, though I haven't tested/validated this.

While these warnings can be ignored/suppressed on an ad-hoc basis, they create noise. Static analysis tools will not warn about such unused parameters when a method is implementing a contract using an abstract method or an interface, but as neither use case applies here, the warning fires.

Omitting the parameter breaks the contract for custom binding with Minimal APIs, so it cannot be removed without stopping the feature from working.

### Describe the solution you'd like

Minimal APIs considers the `ParameterInfo` to be "optional" (not an optional parameter as in the formal C# language specification), so that if the parameter is not needed by the developer/implementation, it can be omitted and the custom binding behaviour will still work.

```csharp
public static ValueTask<PagingData?> BindAsync(HttpContext context)
```

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/36415

Prefers `BindAsync(HttpContext, ParameterInfo)` over `BindAsync(HttpContext)`.

### CODE REVIEW

File: src/Shared/ParameterBindingMethodCache.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public bool HasTryParseMethod(ParameterInfo parameter)
         }
 
         public bool HasBindAsyncMethod(ParameterInfo parameter) =>
-            FindBindAsyncMethod(parameter) is not null;
+            FindBindAsyncMethod(parameter).Item1 is not null;
```
Comment (davidfowl) [inline src/Shared/ParameterBindingMethodCache.cs:50]: I dislike the use of Item1 here
```suggestion
            var (expression, _) = FindBindAsyncMethod(parameter);
            return expression is not null;
```

File: src/Shared/ParameterBindingMethodCache.cs [hunk 2]
```diff
@@ -47,7 +47,8 @@ public bool HasTryParseMethod(ParameterInfo parameter)
         }
 
         public bool HasBindAsyncMethod(ParameterInfo parameter) =>
```
Comment (BrennanConroy) [inline src/Shared/ParameterBindingMethodCache.cs:49]:
```suggestion
        public bool HasBindAsyncMethod(ParameterInfo parameter)
```

File: src/Shared/ParameterBindingMethodCache.cs [hunk 3]
```diff
@@ -47,7 +47,8 @@ public bool HasTryParseMethod(ParameterInfo parameter)
...
-            FindBindAsyncMethod(parameter) is not null;
+            var (expression, _) = FindBindAsyncMethod(parameter);
```
Comment (BrennanConroy) [inline src/Shared/ParameterBindingMethodCache.cs:50]:
```suggestion
        {
            var (expression, _) = FindBindAsyncMethod(parameter);
```

File: src/Shared/ParameterBindingMethodCache.cs [hunk 4]
```diff
@@ -47,7 +47,8 @@ public bool HasTryParseMethod(ParameterInfo parameter)
...
+            return expression is not null;
```
Comment (BrennanConroy) [inline src/Shared/ParameterBindingMethodCache.cs:51]:
```suggestion
            return expression is not null;
        }
```

File: src/Shared/ParameterBindingMethodCache.cs [hunk 5]
```diff
@@ -128,12 +131,18 @@ public bool HasTryParseMethod(ParameterInfo parameter)
             return _stringMethodCallCache.GetOrAdd(type, Finder);
         }
 
-        public Expression? FindBindAsyncMethod(ParameterInfo parameter)
+        public (Expression?, int) FindBindAsyncMethod(ParameterInfo parameter)
```
Comment (BrennanConroy) [inline src/Shared/ParameterBindingMethodCache.cs:134]:
```suggestion
        public (Expression? Expression, int ParamCount) FindBindAsyncMethod(ParameterInfo parameter)
```
Would this be better @davidfowl ?

Comment (davidfowl) [reply src/Shared/ParameterBindingMethodCache.cs:134]: I don't think this changed much but it's fine

**Review by halter73 — APPROVED** [review]
Nice test coverage!

**Review by davidfowl — APPROVED** [review]

**Pull request merged**